### PR TITLE
[enhancement](exception) Column filter/replicate supports exception safety

### PR DESCRIPTION
**Pull request by xinyiZzz**
# Proposed changes

Issue Number: close #xxx

## Problem summary

column filter/replicate supports exception safety, which will actively cancel query when an exception is thrown.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/runtime/thread_context.h [hunk 1]
```diff
@@ -312,33 +312,22 @@ class StopCheckThreadMemTrackerLimit {
...
-                        doris::thread_context()->thread_mem_tracker_mgr->exceed_mem_limit_msg())); \
-            }                                                                                      \
-        }                                                                                          \
+#define RETURN_IF_CATCH_EXCEPTION(stmt)                                                        \
+    do {                                                                                       \
+        try {                                                                                  \
+            doris::thread_context()->thread_mem_tracker_mgr->clear_exceed_mem_limit_msg();     \
+            enable_thread_catch_bad_alloc++;                                                   \
+            { stmt; }                                                                          \
+            enable_thread_catch_bad_alloc--;                                                   \
```
Comment (yiguolei) [inline be/src/runtime/thread_context.h:321]: use defer to --, if returned in line 320

Comment (xinyiZzz) [reply be/src/runtime/thread_context.h]: fixed

**Comment by xinyiZzz** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.37 seconds
 stream load tsv:          434 seconds loaded 74807831229 Bytes, about 164 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          72 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230411140754_clickbench_pr_127640.html

File: be/src/vec/core/block.h [hunk 2]
```diff
@@ -264,9 +265,11 @@ class Block {
 
     void append_block_by_selector(MutableBlock* dst, const IColumn::Selector& selector) const;
 
+    // need exception safety
     static void filter_block_internal(Block* block, const std::vector<uint32_t>& columns_to_filter,
```
Comment (yiguolei) [inline be/src/vec/core/block.h:269]: it's better to modify this method to be exception safe, and return error status.
then we only catch exception inside this method. The callers only need to return_if_error

File: be/src/runtime/thread_context.h [hunk 3]
```diff
@@ -312,33 +312,22 @@ class StopCheckThreadMemTrackerLimit {
...
-                        "PreCatch {}, {}", e.what(),                                               \
-                        doris::thread_context()->thread_mem_tracker_mgr->exceed_mem_limit_msg())); \
-            }                                                                                      \
-        }                                                                                          \
+#define RETURN_IF_CATCH_EXCEPTION(stmt)                                                        \
+    do {                                                                                       \
+        try {                                                                                  \
+            doris::thread_context()->thread_mem_tracker_mgr->clear_exceed_mem_limit_msg();     \
+            doris::enable_thread_catch_bad_alloc++;                                            \
+            Defer defer {[&]() { doris::enable_thread_catch_bad_alloc--; }};                   \
```
Comment (yiguolei) [inline be/src/runtime/thread_context.h:320]: why do we need such variable "enable_thread_catch_bad_alloc" ?

File: be/src/vec/core/block.cpp [hunk 4]
```diff
@@ -738,7 +738,7 @@ Status Block::filter_block(Block* block, const std::vector<uint32_t>& columns_to
         for (size_t i = 0; i < size; ++i) {
             filter_data[i] &= !null_map[i];
         }
```
Comment (yiguolei) [inline be/src/vec/core/block.cpp:740]: add [[nodiscard]]  to filter_block method to force the caller check the status.

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**